File: Windows/Type4Me/LLM/ClaudeChatClient.cs
using System.IO;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using Type4Me.Models;
using Type4Me.Services;

namespace Type4Me.LLM;

/// <summary>
/// Anthropic Messages API streaming LLM client.
/// </summary>
public sealed class ClaudeChatClient : ILLMClient
{
    private static readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(30) };

    public async Task WarmUpAsync(string baseURL)
    {
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, baseURL);
            await _http.SendAsync(req);
            DebugFileLogger.Log($"[Claude] Connection pre-warmed to {baseURL}");
        }
        catch { /* ignore */ }
    }

    public async Task<string> ProcessAsync(string text, string prompt, LLMConfig config, CancellationToken ct = default)
    {
        var trimmedText = text.Trim();
        if (string.IsNullOrEmpty(trimmedText)) return text;

        var finalPrompt = prompt.Replace("{text}", trimmedText);
        var url = $"{config.BaseURL.TrimEnd('/')}/messages";

        using var request = new HttpRequestMessage(HttpMethod.Post, url);
        request.Headers.Add("x-api-key", config.ApiKey);
        request.Headers.Add("anthropic-version", "2023-06-01");
        request.Headers.Add("Accept", "text/event-stream");

        var body = new
        {
            model = config.Model,
            max_tokens = 4096,
            messages = new[] { new { role = "user", content = finalPrompt } },
            stream = true,
        };

        request.Content = new StringContent(
            JsonSerializer.Serialize(body),
            Encoding.UTF8,
            "application/json");

        DebugFileLogger.Log($"[Claude] Request: {text.Length} chars, model={config.Model}");

        using var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct);

        if (!response.IsSuccessStatusCode)
        {
            DebugFileLogger.Log($"[Claude] HTTP {(int)response.StatusCode}");
            throw new HttpRequestException($"Claude request failed with status {(int)response.StatusCode}");
        }

        // Parse SSE stream (Anthropic format)
        using var stream = await response.Content.ReadAsStreamAsync(ct);
        using var reader = new StreamReader(stream);

        var result = new StringBuilder();

        while (await reader.ReadLineAsync(ct) is { } line)
        {
            if (!line.StartsWith("data: ")) continue;
            var payload = line[6..];

            try
            {
                using var doc = JsonDocument.Parse(payload);
                var root = doc.RootElement;
                var type = root.TryGetProperty("type", out var t) ? t.GetString() : null;

                switch (type)
                {
                    case "content_block_delta":
                        if (root.TryGetProperty("delta", out var delta) &&
                            delta.TryGetProperty("text", out var textEl))
                        {
                            var chunk = textEl.GetString();
                            if (chunk != null) result.Append(chunk);
                        }
                        break;

                    case "message_stop":
                        goto done;
                }
            }
            catch { /* skip malformed chunks */ }
        }

        done:
        DebugFileLogger.Log($"[Claude] Result: {result.Length} chars");
        return result.ToString();
    }
}

File: MacOS/Type4Me/LLM/ClaudeChatClient.swift
import Foundation
import os

actor ClaudeChatClient: LLMClient {

    private let logger = Logger(subsystem: "com.type4me.llm", category: "ClaudeChatClient")

    /// Pre-establish TCP+TLS connection so the first real request skips handshake.
    func warmUp(baseURL: String) async {
        guard let url = URL(string: baseURL) else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 5
        _ = try? await URLSession.shared.data(for: request)
        logger.info("Claude connection pre-warmed to \(baseURL)")
    }

    /// Process text through Anthropic Messages API (streaming).
    func process(text: String, prompt: String, config: LLMConfig) async throws -> String {
        let trimmedText = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedText.isEmpty else { return text }
        let finalPrompt = prompt.replacingOccurrences(of: "{text}", with: trimmedText)

        guard let url = URL(string: "\(config.baseURL)/messages") else {
            throw LLMError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(config.apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 30

        let body = ClaudeRequest(
            model: config.model,
            max_tokens: 4096,
            system: nil,
            messages: [ClaudeMessage(role: "user", content: finalPrompt)],
            stream: true
        )
        request.httpBody = try JSONEncoder().encode(body)

        logger.info("Claude request: \(text.count) chars, model=\(config.model)")

        let (bytes, response) = try await URLSession.shared.bytes(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw LLMError.requestFailed(0)
        }
        guard http.statusCode == 200 else {
            logger.error("Claude HTTP \(http.statusCode)")
            throw LLMError.requestFailed(http.statusCode)
        }

        // Parse SSE stream (Anthropic format)
        var result = ""
        for try await line in bytes.lines {
            guard line.hasPrefix("data: ") else { continue }
            let payload = String(line.dropFirst(6))
            guard let data = payload.data(using: .utf8),
                  let event = try? JSONDecoder().decode(ClaudeStreamEvent.self, from: data)
            else { continue }

            switch event.type {
            case "content_block_delta":
                if let delta = event.delta, let text = delta.text {
                    result += text
                }
            case "message_stop":
                break
            default:
                continue
            }
        }

        logger.info("Claude result: \(result.count) chars")
        return result
    }
}

// MARK: - Request Types

private struct ClaudeRequest: Encodable, Sendable {
    let model: String
    let max_tokens: Int
    let system: String?
    let messages: [ClaudeMessage]
    let stream: Bool
}

private struct ClaudeMessage: Encodable, Sendable {
    let role: String
    let content: String
}

// MARK: - Stream Response Types

private struct ClaudeStreamEvent: Decodable, Sendable {
    let type: String
    let delta: ClaudeDelta?
}

private struct ClaudeDelta: Decodable, Sendable {
    let text: String?
}
